### ISSUE TITLE
fuge test

### ISSUE THREAD

**Issue by davidmarkclements** [apparatus/fuge#1]
from the compose yaml file, fuge knows where each service is - and it also knows how to run tests from the package.json file

what would be nice though is to have a way to aggregate all of the results and coverage centrally, 

the https://www.npmjs.com/package/nyc tool sits on top of istanbul and checks coverage over multiple processes and then generates a single report

tape output could also be easily aggregated - as long as there's a way (by some convention) to get the TAP output of each test a user could simply run

```
fuge test
```

to test an entire deployment in a central place

**Comment by mcollina** [comment]
That would be pretty awesome.

I would also say that we can support integration tests in this way, which is something extremely useful and hard to achieve.

**Comment by davidmarkclements** [comment]
yeah - if we can enforce a testing convention for each service (site/api too) then this should be trivial.

rules would be

* each service must use TAP
* each service must supply TAP output
  * e.g. if npm test script pipes through tap-dot then we can't use that for fuge test
  * EITHER we have a convention where fuge runs all test folders (which is actually tighter) 
  * OR npm test MUST supply TAP output, and if you want nice individual report its npm run test-pretty or something
  * OR by convention an additional script "test-raw" or "fuge-test" or "test-tap" or something, is supplied that throws tap output, then `fuge test` spawns `npm run test-raw`.

**Comment by mcollina** [comment]
The command to run the tests (producing the TAP output) should be in the yml file.

**Comment by bahmutov** [comment]
Or just leave the test command as `npm test` but require TAP output?

**Comment by mcollina** [comment]
I guess we can default it to `npm test` and require TAP output. But it would require config at project level.

**Comment by davidmarkclements** [comment]
why not both 

1) default to running "npm test"
2) allow test command override in yaml files for special 

someone way wish to develop in a  way where npm test outputs the report instead
of the raw TAP, and have `npm run test-tap`  or maybe `npm test -- --tap` for the tap
output - in this case they can override it in the yaml file

**Comment by AdrianRossouw** [comment]
just an fyi: seneca has standardised on lab for all services, and any seneca generators we end up with will probably be required to use lab as it's testing environment.

I don't think it's feasible for us to require services to output tap (even though I also prefer it)

**Comment by davidmarkclements** [comment]
Since we control lab - we can make it output tap right? Mocha, for instance, can use a tap reporter. @geek - thoughts on this?

Whatever we use, the idea is to stream test results in a standard format to a central point

**Comment by geek** [comment]
@davidmarkclements absolutely:  `lab -r tap` works today

**Comment by davidmarkclements** [comment]
@geek any possibility of making that the default for fuge compat?

**Comment by geek** [comment]
@davidmarkclements :/ unfortunately not. The default is `console`, which is what most people prefer. When generating the package.json test script entry is there a reason we can't add `-r tap` ?

**Comment by AdrianRossouw** [comment]
i didn't know that it exposed tap, which is cool.

how about we add an npm run test:tap as our standard.

**Comment by davidmarkclements** [comment]
@geek sure we can do that - would have been cool if we could get tool defaults working together - obviously it would be a major (as in semver major) change for lab, but if it worked similar to `tap` and `tape` (where TAP is the default but to get human readable output you pipe, e.g. "lab tests | lab-console" ) that would be (imo) awesome

if not we can take the args approach but that will mean that if someone moves a seneca deployment into fuge, and they want multi-service test output they'll need to change the test arguments for each service (unless we maybe have a config option in fuge that passes an argument to npm test, e.g. `npm test -- -r tap`) 

on a side note - how does `lab` do with coverage (e.g. does it include `nyc` by default for multi-process coverage?)

**Comment by AdrianRossouw** [comment]
that assumes that the service is using lab.

I think that this functionality is cool enough to ask people to modify their package.json.
btw, what's to say there is even a package.json. or npm doing the testing.

the test command should be configurable inside fuge, so you wouldn't have to modify your project.
(i think it is already).

**Comment by davidmarkclements** [comment]
How about the following

1) control test command per service to compose yaml file 
2) default to npm test 
3) make default (npm test) configurable from config.js
4) when initialising the test, then moment we find were not getting tap - exit the test and reattempt with `npm test -- -r tap` also issue some kind of warning with link to explanation

**Comment by AdrianRossouw** [comment]
i think I understand what you mean, as long as this sentence should say "in compose yaml file"

> control test command per service __in__ compose yaml file

**Comment by davidmarkclements** [comment]
yes - in
